### PR TITLE
Add T object and Test method to Mock

### LINKED ISSUE #489: Should mock really panic instead of t.Fail?
I'm not sure which is best overall, but I'd like to share my experience just to bring this into consideration.

I was testing a component of my software which will recover whenever a panic occurs, so having an unexpected method call panic broke the logic of my test.

If mock was passed the testing.T object and called t.Fail instead of panic, my test would work fine. That said, I can also imagine circumstances where you'd rather have a panic.

It might be good to make this behaviour configurable. The way I would do this is by adding an exported field to the mock.Mock struct `T *testing.T`, then if `T != nil` call `panic`, otherwise `T.Error`

### PR DESCRIPTION
**Pull request by posener**
This makes is possible to fail the test instead of panicing in case
that the method was called with unexpected arguments.

Fixes #489

### CODE REVIEW

File: mock/mock.go [hunk 1]
```diff
@@ -191,6 +191,10 @@ type Mock struct {
 	// Holds the calls that were made to this mocked object.
 	Calls []Call
 
+	// T is An optional variable that holds the test struct, to be used when an
+	// invalid mock call was made.
+	T TestingT
```
Comment (ernesto-jimenez) [inline mock/mock.go]: Can you unexport `TestingT`?

The package already exports too much of it's internals and we should keep new things to a minimum.

File: mock/mock.go [hunk 2]
```diff
@@ -501,7 +526,7 @@ type Arguments []interface{}
 const (
 	// Anything is used in Diff and Assert when the argument being tested
 	// shouldn't be taken into consideration.
-	Anything string = "mock.Anything"
+	Anything = "mock.Anything"
```
Comment (ernesto-jimenez) [inline mock/mock.go]: FYI, for future contributions, it would be best to keep this kind of things in a different PR.

File: mock/mock_test.go [hunk 3]
```diff
@@ -92,6 +92,34 @@ func (i *TestExampleImplementation) TheExampleMethodFuncType(fn ExampleFuncType)
 	return args.Error(0)
 }
 
+// MockTestingT mocks a test struct
+type MockTestingT struct {
+	Mock
```
Comment (ernesto-jimenez) [inline mock/mock_test.go]: I would rather avoid using `mock` to mock a dependency for its own test :)

Let's implement this with a simple mock instead to keep the tests for the package depending on the package itself working correctly :)

File: mock/mock.go (hunk 1 above)
Comment (posener) [reply mock/mock.go]: I didn't add the `TestingT` interface you mean unexport `T` I guess?

**Comment by posener** [comment]
Thanks @ernesto-jimenez , pushed the fixes you asked

**Comment by posener** [comment]
I'm closing this (not deleting, so you can still merge)
I think I have a nicer API, I'll open a new PR, you can merge the one you like better..

**Comment by posener** [comment]
Sorry, the new API didn't work out for me.
I think that once this is merged, we could add a `New` function in mockery to make the creation of the mock and assigning the test field in one command.

**Pull request closed without merging**